pkg/app/piped/apistore/eventstore/store.go

```diff
@@ -105,104 +102,53 @@ func (s *store) Run(ctx context.Context) error {
 // sync fetches a list of events newly created after its own milestone,
 // and updates the cache of latest events.
 func (s *store) sync(ctx context.Context) error {
+	// Fetch events in descending order.
 	resp, err := s.apiClient.ListEvents(ctx, &pipedservice.ListEventsRequest{
-		From:  s.milestone,
-		Order: pipedservice.ListOrder_ASC,
+		From:   time.Now().Add(-cacheTTL).Unix(),
+		Order:  pipedservice.ListOrder_DESC,
+		Status: pipedservice.ListEventsRequest_NOT_HANDLED,
 	})
 	if err != nil {
 		return fmt.Errorf("failed to list events: %w", err)
 	}
 	if len(resp.Events) == 0 {
+		s.notHandledEvents.Store(make(map[string][]*model.Event))
 		return nil
 	}
 
-	// Eliminate events that have duplicated key.
-	filtered := make(map[string]*model.Event, len(resp.Events))
+	notHandledEvents := make(map[string][]*model.Event)
 	for _, e := range resp.Events {
-		filtered[e.EventKey] = e
-	}
-	// Make the cache up-to-date.
-	s.mu.Lock()
-	for key, event := range filtered {
-		cached, ok := s.latestEvents[key]
-		if ok && cached.CreatedAt > event.CreatedAt {
-			continue
-		}
-		s.latestEvents[key] = event
+		notHandledEvents[e.EventKey] = append(notHandledEvents[e.EventKey], e)
 	}
-	s.mu.Unlock()
+	s.notHandledEvents.Store(notHandledEvents)
 
-	// Set the latest one within the result as the next time's "from".
-	s.milestone = resp.Events[len(resp.Events)-1].CreatedAt + 1
 	return nil
 }
 
-func (s *store) GetLatest(ctx context.Context, name string, labels map[string]string) (*model.Event, bool) {
+func (s *store) ListNotHandled(name string, labels map[string]string, minCreatedAt int64, limit int) []*model.Event {
+	if limit == 0 {
+		return []*model.Event{}
+	}
 	key := model.MakeEventKey(name, labels)
-	s.mu.RLock()
-	event, ok := s.latestEvents[key]
-	s.mu.RUnlock()
-	if ok {
-		return event, true
+	notHandledEvents := s.notHandledEvents.Load()
+	if notHandledEvents == nil {
+		return []*model.Event{}
 	}
 
-	// If not found in the cache, fetch from the control-plane.
-	resp, err := s.apiClient.GetLatestEvent(ctx, &pipedservice.GetLatestEventRequest{
-		Name:   name,
-		Labels: labels,
-	})
-	if status.Code(err) == codes.NotFound {
-		s.logger.Info("event not found in control-plane",
-			zap.String("event-name", name),
-			zap.Any("labels", labels),
-		)
-		return nil, false
-	}
-	if err != nil {
-		s.logger.Error("failed to get the latest event", zap.Error(err))
-		return nil, false
+	events := notHandledEvents.(map[string][]*model.Event)[key]
+	out := make([]*model.Event, 0, len(events))
+	for _, e := range events {
+		if e.CreatedAt < minCreatedAt {
+			break
+		}
+		out = append(out, e)
 	}
-
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	cached, ok := s.latestEvents[key]
-	if ok && cached.CreatedAt > event.CreatedAt {
-		return resp.Event, true
+	if len(out) < limit {
+		return out
 	}
-	s.latestEvents[key] = resp.Event
-	return resp.Event, true
+	return out[:limit]
```

Comment:
nit: We can include the limit check into this loop.

Reply:
Right 👍